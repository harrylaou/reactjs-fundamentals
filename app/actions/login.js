export const LOGIN_USER_OK = 'LOGIN_USER_OK'
export const LOGOUT_USER = 'LOGOUT_USER'

export const loginUser = () => ({
  type: LOGIN_USER_OK
})

export const logoutUser = () => ({
  type: LOGOUT_USER
})