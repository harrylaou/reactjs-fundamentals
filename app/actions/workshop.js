export const RECEIVE_WORKSHOP = 'RECEIVE_WORKSHOP'


export const receiveWorkshop = workshop => ({
  type: RECEIVE_WORKSHOP,
  workshop
})