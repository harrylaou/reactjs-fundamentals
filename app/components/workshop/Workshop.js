import React, {Component} from 'react'
import {
  Card,
  CardActions,
  CardHeader,
  CardMedia,
  CardTitle,
  CardText
} from 'material-ui/Card'
import es6promise from 'es6-promise'
import 'isomorphic-fetch'

es6promise.polyfill()

const Workshop = props => {

  let workshop = props.workshop
  let id = props.id
  console.log("Workshop id", props.id)

  if (workshop && workshop.id != id) {
    props.fetchWorkshop(id)
  }
  let title = ''
  let price = ''
  let url = ''
  let avatar = ''

  if (workshop) {
    title = `${workshop.title}`
    price = `${workshop.price}`
    url = `${workshop.url}`
    avatar = `${workshop.instructors[0].avatar}`
  }

  return (
    <Card>
      <CardHeader title={title} subtitle={url} avatar={avatar}/>
      <CardActions/>
    </Card>
  )
}

export default Workshop
