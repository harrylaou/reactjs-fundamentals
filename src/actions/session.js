export const LOG_OUT = 'LOG_OUT'
export const LOGGED_IN = 'LOGGED_IN'

export const logout = () => ({
  type: LOG_OUT
})

export const loggedIn = token => ({
  type: LOGGED_IN,
  token
})
