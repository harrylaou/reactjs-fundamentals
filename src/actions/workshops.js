export const RECEIVE_WORKSHOPS = 'RECEIVE_WORKSHOPS'

export const receiveWorkshops = workshops => ({
  type: RECEIVE_WORKSHOPS,
  workshops
})
