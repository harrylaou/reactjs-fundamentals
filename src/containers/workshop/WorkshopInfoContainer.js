import React, {Component} from 'react'
import {connect} from 'react-redux'
import WorkshopInfo from '../../components/workshop/WorkshopInfo'
import * as api from '../../api'
import * as actions from '../../actions/workshop'

class WorkshopInfoContainer extends Component {
  constructor() {
    super()
    this.fetchWorkshop = this.fetchWorkshop.bind(this)
  }

  componentDidMount() {
    this.fetchWorkshop(this.props.params.workshop)
  }

  componentWillReceiveProps(nextProps) {
    // If you are thinking that performance could be improved by caching
    // the user, we agree. But we want to keep this example simple.
    // We are going to improve this code in the next lessons. We are building up :)
    if (this.props.params.workshop !== nextProps.params.workshop) {
      this.fetchWorkshop(nextProps.params.workshop)
    }
  }

  // Imagine we want to fetch another workshop from another component. How do you
  // think we could reuse this code?
  fetchWorkshop(workshop) {
    api
      .getWorkshop(workshop)
      .then((workshop) => {
        this.props.dispatch(actions.receiveWorkshop(workshop))
      })
      .catch((err)=> {
        console.log(err)
      })
  }

  render() {
    return (
      <WorkshopInfo
        {...this.props}
        workshop={this.props.workshop}
      />
    )
  }
}

WorkshopInfoContainer.propTypes = {
  params: React.PropTypes.object.isRequired,
  dispatch: React.PropTypes.func.isRequired,
  workshop: React.PropTypes.object
}

const mapStateToProps = (state) => ({
  workshop: state.workshop
})

// Could you refactor this to make it better? Hint: LoginContainer
const mapDispatchToProps = (dispatch) => ({
  dispatch
})

// Do you think the order of these components matter?
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(WorkshopInfoContainer)
