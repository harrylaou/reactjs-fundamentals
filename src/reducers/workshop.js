import { RECEIVE_WORKSHOP } from '../actions/workshop'

const workshop = (state = null, action) => {
  switch (action.type) {
    case RECEIVE_WORKSHOP: {
      return action.workshop
    }
    default:
      return state
  }
}

export default workshop
