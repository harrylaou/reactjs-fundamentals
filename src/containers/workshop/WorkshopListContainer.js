import React, {Component} from 'react'
import {connect} from 'react-redux'
import WorkshopList from '../../components/workshop/WorkshopList'
import * as api from '../../api'
import * as actions from '../../actions/workshops'
import withWidth from '../../utils/WithWidth'
import { withRouter } from 'react-router'

class WorkshopListContainer extends Component {
  constructor() {
    super()
    this.showWorkshop = this.showWorkshop.bind(this)
  }

  componentDidMount() {
    // Imagine we want to fetch workshops from another component. How do you
    // think we could reuse this code?
    api
      .getWorkshops()
      .then((workshops) => {
        this.props.dispatch(actions.receiveWorkshops(workshops))
      })
      .catch((err)=> {
        console.log(err)
      })
  }

  showWorkshop(workshop) {
    this.props.router.push(`/workshops/${workshop.id}`)
  }

  render() {
    return (
      <WorkshopList
        {...this.props}
        workshop={this.props.params.workshop}
        workshops={this.props.workshops}
        showWorkshop={this.showWorkshop}
        width={this.props.width}
      />
    )
  }
}

WorkshopListContainer.propTypes = {
  width: React.PropTypes.number.isRequired,
  router: React.PropTypes.object.isRequired,
  workshops: React.PropTypes.array,
  dispatch: React.PropTypes.func.isRequired,
  params: React.PropTypes.object.isRequired
}

const mapStateToProps = (state) => ({
  workshops: state.workshops
})

// Could you refactor this to make it better? Hint: LoginContainer
const mapDispatchToProps = (dispatch) => ({
  dispatch
})

// Do you think the order of these components matter?
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(withWidth(WorkshopListContainer)))
