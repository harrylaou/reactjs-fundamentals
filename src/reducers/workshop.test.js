import workshop from './workshop'
import {RECEIVE_WORKSHOP} from '../actions/workshop'

describe('#Reducer: Workshop', () => {
  it('Return default state',() => {
    expect(workshop(undefined, { type: 'default_action' })).toEqual(null)
    expect(workshop(undefined, { type: 'default_action' })).toMatchSnapshot()
  })

  it('Return state',() => {
    expect(workshop("data", { type: 'another_action' })).toEqual("data")
    expect(workshop("data", { type: 'another_action' })).toMatchSnapshot()
  })

  it(`When Action ${RECEIVE_WORKSHOP} return state`,() => {
    expect(workshop("", { type: RECEIVE_WORKSHOP, workshop: "React Fundamentals" })).toEqual("React Fundamentals")
    expect(workshop("", { type: RECEIVE_WORKSHOP, workshop: "React Fundamentals" })).toMatchSnapshot()
  })
})