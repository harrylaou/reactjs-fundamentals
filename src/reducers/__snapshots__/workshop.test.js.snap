// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#Reducer: Workshop Return default state 1`] = `null`;

exports[`#Reducer: Workshop Return state 1`] = `"data"`;

exports[`#Reducer: Workshop When Action RECEIVE_WORKSHOP return state 1`] = `"React Fundamentals"`;
